backend/editor/controllers/search_controller.py

```diff
@@ -0,0 +1,374 @@
+
+    The possible filters are:
+      - `is`: `root` is the only possible value. It allows to filter on the root nodes.
+        It cannot be negated.
+      - `language`: the value is a language code. It allows to filter on if the language exists
+        on the node. It can be negated.
+      - `parent`: the value is a node's id. It allows to filter on if the node is a
+        parent of the node with the given id. It cannot be negated.
+      - `child`: the value is a node's id. It allows to filter on if the node is a child of
+        the node with the given id. It cannot be negated.
+      - `ancestor`: the value is a node's id. It allows to filter on if the node is an ancestor
+        of the node with the given id. It cannot be negated.
+      - `descendant`: the value is a node's id. It allows to filter on if the node is a descendant
+        of the node with the given id. It cannot be negated.
+
+    Examples:
+    - "is:root language:en not(language):fr"
+    - "parent:"en:apple juice" descendant:en:juices "fruit concentrate""
+    """
+
+    search_terms = split_query_into_search_terms(query)
+
+    name_search_terms = []
+    filter_search_terms = []
+
+    for search_term in search_terms:
+        if (filter_search_term := parse_filter_search_term(search_term)) is not None:
+            filter_search_terms.append(filter_search_term)
+        else:
+            name_search_terms.append(search_term)
+
+    return Query(project_id, name_search_terms, filter_search_terms)
+
+
+def build_lucene_name_search_query(search_value: str) -> str | None:
+    """
+    The name search term can trigger two types of searches:
+    - if the search value is in the format `language_code:raw_search_value`,
+      it triggers a search on the tags_ids_{language_code} index
+    - else it triggers a search on the tags_ids index
+
+    If the `raw_search_value` is surrounded by quotes, the search will be exact.
+    Otherwise, the search is fuzzy when the search value is longer than 4 characters
+    (the edit distance depends of the length of the search value)
+    """
+    language_code = None
+
+    if len(search_value) > 2 and search_value[2] == ":" and search_value[0:2].isalpha():
+        language_code, search_value = search_value.split(":", maxsplit=1)
+        language_code = language_code.lower()
+
+    def get_search_query() -> str | None:
+        if search_value.startswith('"') and search_value.endswith('"'):
+            return search_value if len(search_value) > 2 else None
+
+        if language_code is not None:
+            normalized_text = parser_utils.normalize_text(search_value, language_code)
+        else:
+            normalized_text = parser_utils.normalize_text(search_value)
+
+        # If normalized text is empty, no searches are found
+        if normalized_text.strip() == "":
+            return None
+
+        tokens = normalized_text.split("-")
+
+        def get_token_query(token: str) -> str:
```

Comment:
a function within a function within a function ?

Why isn't it just a function outside this scope ?(normally you use function within function if it depends on many variables)

IMO, even for get_search_query I don't see a necessity for an inner function. You can make a private function (_get_search_query) to make it clear it's only locally useful.